dns-policy-operator/src/charm.py

```diff
@@ -0,0 +1,208 @@
+            self.unit.status = ops.MaintenanceStatus("Workload not yet ready.")
+            return
+
+        if not self._database.is_relation_ready():
+            self.unit.status = ops.WaitingStatus("Waiting for a database integration.")
+            return
+
+        self.unit.status = ops.ActiveStatus()
+
+    def _on_reconcile(self, _: ReconcileEvent) -> None:
+        if not self.model.unit.is_leader():
+            return
+        relation_data = self.dns_record_provider.get_remote_relation_data()
+        if relation_data is None:
+            logger.debug("Reconciliation: no provider relation data found")
+            return
+        entries: list[models.DnsEntry] = []
+        for record_requirer_data, _ in relation_data:
+            for entry in record_requirer_data.dns_entries:
+                entries.append(entry)
+        if not entries:
+            logger.debug("Reconciliation: no entry found in provider data")
+            return
+
+        token = self.dns_policy.get_api_root_token()
+        self.dns_policy.send_requests(token, entries)
+
+        approved_requests = self.dns_policy.get_approved_requests(token)
+        dns_record_requirer_data = dns_record.DNSRecordRequirerData(dns_entries=approved_requests)
+        for relation in self.model.relations[self.dns_record_requirer.relation_name]:
+            self.dns_record_requirer.update_relation_data(relation, dns_record_requirer_data)
+
+    def dns_record_relations_data_to_entries(
+        self,
+        relation_data: list[
+            tuple[dns_record.DNSRecordRequirerData, dns_record.DNSRecordProviderData]
+        ],
+    ) -> list[models.DnsEntry]:
+        """Convert DNSRecordRequirerData to a list of DnsEntry.
+
+        Args:
+            relation_data: input relation data
+        Returns:
+            A list of DnsEntry
+        """
+        entries: list[models.DnsEntry] = []
+        for record_requirer_data, _ in relation_data:
+            for entry in record_requirer_data.dns_entries:
+                entries.append(models.create_dns_entry_from_requirer_entry(entry))
+        return entries
+
+    def _on_config_changed(self, _: ops.ConfigChangedEvent) -> None:
+        """Handle changed configuration."""
+        self.unit.status = ops.MaintenanceStatus("Configuring workload")
+        self.dns_policy.configure(
+            {
+                "debug": "true" if self.config["debug"] else "false",
+                "allowed-hosts": json.dumps(
+                    [e.strip() for e in str(self.config["allowed-hosts"]).split(",")]
+                ),
+            }
+        )
+
+    def _on_start(self, _: ops.StartEvent) -> None:
+        """Handle start event."""
+
+    def _on_install(self, _: ops.InstallEvent) -> None:
+        """Handle install event."""
+        self.unit.status = ops.MaintenanceStatus("Preparing dns-policy-app")
+        self.dns_policy.setup(self.unit.name)
+        self._timer.start(self.unit.name, "reconcile", "30s", "1m")
```

Comment:
It would be helpful to add this information as a log in case of troubleshooting.

Reply:
I added a line of log when the timer starts